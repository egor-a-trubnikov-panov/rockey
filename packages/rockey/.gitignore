/etc

/tasks/currentResults.json

/css
/plugins
/styleSheets
/utils


/condition.js
/index.js
/insert.js
/rule.js
/speedy.js
/when.js
/rockey.min.js
