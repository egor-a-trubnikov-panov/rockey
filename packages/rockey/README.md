# Rockey

Stressless styles for components using js. Write CSS with functional mixins according to the component structure using components names.

<img src="http://i.imgur.com/ULoeOL4.png" height="16"/> [Medium: CSS in JS. Rockey](https://medium.com/@valeriy.sorokobatko/)

![Rockey tests](https://api.travis-ci.org/tuchk4/rockey.svg?branch=master) ![rockey gzip size](http://img.badgesize.io/https://unpkg.com/rockey@0.0.5-alpha.32af5f74/rockey.min.js?compression=gzip&label=rockey%20gzip)

```bash
npm install --save rockey
```

## Api

- [rule](#rule)
  - [rule.getClassList](#rule.getClassList)
  - [rule.addParent](#rule.addParent)
  - [rule.wrapWith](#rule.wrapWith)
  - [rule.addMixins](#rule.addMixins)
  - [rule.transform](#rule.transform)
- [when](#when)
- [condition](#condition)
- [insert](#insert)


### rule

```js
import rule from 'rockey';

const css = rule`
  Button {
    color: black;

    Icon {
      margin: 5px;
    }
  }
`;
```

> NOTE: that in most cases developers uses only *getClassList* and *addParent* methods.
Other methods are mostly used in internal rockey infrastructure, for integrations with
other libraries and for new features.

#### rule.getClassList

```js
import rule from 'rockey';

const css = rule`
  Button {
    color: black;

    Icon {
      margin: 5px;
    }
  }
`;

const classList = css.getClassList();
const className = classList.Button;
```

Inserted CSS:

```css
.Button-{{ hash }}
  color: black;
}

.Button-{{ hash }} .Icon-{{ hash }} {
  margin: 5px;
}
```

#### rule.addParent

Add parent rule. Child rule classnames will be merged with parent rule classnames.
It is possible to add only one parent.

```js
import rule from 'rockey';

const baseButtonCss = rule`
  Button {
    color: black;
    background: white;

    ${props => props.small ? `font-size: 12px;` : null}
  }
`;

const primaryButtonCss = rule`
  PrimaryButton {
    color: blue;
  }
`;

primaryButtonCss.addParent(baseButtonCss);

primaryButtonCss.getClassList({
  small: true
});

primaryButtonCss.PrimaryButton // ['PrimaryButton-{{ hash }}', 'Button-{{ hash }}', 'AnonMixin1-{{ hash }}']
```

Inserted CSS:

```css
.Button-{{ hash }{
  color: black;
  background: white;
}

.PrimaryButton-{{ hash }} {
  color: black;
  background: white;
}

.AnonMixin1-{{ hash }}.Button-{{ hash }} {
  font-size: 12px;
}
```

#### rule.wrapWith

Wrap current rule with another selector.

```js
import rule from 'rockey';

const css = rule`
  Button {
    color: black;
  }
`;

css.wrapWith('Layer');

const classList = css.getClassList();
classList.Layer // ['Layer-{{ hash }}']
```

Inserted CSS:

```css
.Layer-{{ hash }} .Button-{{ hash }} {
  color: black;
}
```

#### rule.addMixins

Api to add mixins. After this *addMixins()* rule should be wrapped with *wrapWith()*

> NOTE: this method is mostly used in internal rockey infrastructure and for integrations with
other libraries. Try not to use it in applications.

```js
import rule from 'rockey';

const css = rule`
  color: black;
`;

css.addMixins([
  props => props.small ? `font-size: 12px;` : null
]);

css.wrapWith('Button');

css.getClassList({
  small: true
});
```

#### rule.transform

Transform current css tree and setup relations between defined components

> NOTE: this method is mostly used in internal rockey infrastructure and for integrations with
other libraries. Try not to use it in applications.

For example: split rule into multiple rules and make each component to be child of previous component.
This is how [rockey-react](https://github.com/tuchk4/rockey/tree/master/packages/rockey-react) *look* feature works.

```js
import rule from 'rockey';

const css = rule`
  Base {
    background: none;
  }

  Sized {
    font-size: 15px;
  }  

  Colored {
    color: red;
  }
`;

const components = css.transform((tree, create) => {
  const rules = {};
  const components = Object.keys(tree.components);
  for (let i = 0; i < components.length; i++) {
    const displayName = components[i];
    const rule = create(tree.components[displayName]);

    if (i) {
      const parentName = components[i - 1]
      rule.addParent(rules[parentName]);
    }

    rules[displayName] = rule;
  }

  return rules;
});

components.Base.getClassList(); // [ "Base-{{ hash }}" ]
components.Sized.getClassList(); // [ "Sized-{{ hash }}", "Base-{{ hash }}" ]
components.Colored.getClassList(); // [ "Colored-{{ hash }}", "Sized-{{ hash }}", "Base-{{ hash }}" ]
```

### when

```js
import when from 'rockey/when';
import { when } from 'rockey';
```

Helps to to keep syntax much better and cleaner when use mixins.
Fulfilled each time when *getClassList()* is called.

```js
import rule from 'rockey';

const css = rule`
  Button {
    color: black;

    ${rockey.when(props => props.primary)`
      color: blue;
    `}

    Icon {
      margin: 5px;
    }
  }
`;

const classList = css.getClassList({
  primary: true
});

classList.Button; // ['Button-{{ hash }}', 'AnonMixin1-{{ hash }}']
```

Inserted CSS:

```css
.Button-{{ hash }} {
  color: black;
}

.AnonMixin1-{{ hash }}.Button-{{ hash }} {
  color: blue;
}

.Button-{{ hash }} .Icon-{{ hash }} {
  font-size: 12px;
}
```

To provide more readable mixin class name - define mixin name at first argument.

```js
import rule from 'rockey';

const css = rule`
  Button {
    color: black;

    ${rockey.when('isPrimary', props => props.primary)`
      color: blue;
    `}

    Icon {
      margin: 5px;
    }
  }
`;

const classList = css.getClassList({
  primary: true
});

classList.Button; // ['Button-{{ hash }}', 'isPrimary-{{ hash }}']
```

Inserted CSS:

```css
.Button-{{ hash }} {
  color: black;
}

.isPrimary-{{ hash }}.Button-{{ hash }} {
  color: blue;
}

.Button-{{ hash }} .Icon-{{ hash }} {
  font-size: 12px;
}
```

Access to props inside CSS rules:

```js
import rule from 'rockey';

const css = rule`
  Button {
    color: black;

    ${rockey.when(props => props.color)(props => `
      color: ${props.color};
    `)}

    Icon {
      margin: 5px;
    }
  }
`;
```

### condition

```js
import condition from 'rockey/condition';
import { condition } from 'rockey';
```

Same as *when* but it is called only when CSS string is parsing.

```js
import rule from 'rockey';

const css = rule`
  Button {
    color: black;

    ${condition(isMobile())`
      font-size: 10px;
    `}

    Icon {
      margin: 5px;
    }
  }
`;
```

### insert

```js
import { insert } from 'rockey';
import insert from 'rockey/insert';
```

Insert CSS rules to the DOM as is. Without any processing.

```js
import rule from 'rockey';

rule.insert`
  body {
    overflow: hidden;
  }

  #hardcode {
    color: red;
  }

  span.my {
    color: green;
  }
`;
```

## Feedback wanted

This is a very new approach and library and not all features are implemented yet. Feel free to [file issue or suggest feature](https://github.com/tuchk4/rockey/issues/new) to help me to make rockey better.
Or ping me on twitter @tuchk4.

ðŸŽ‰

Upcoming plans:

- Make disadvantages list as shorter as possible
- Medium post *"Rockey Under the Hood"*. Topic about how rockey worksâ€Šâ€”â€Šhow to make batch CSS rules insert, how to parse and auto optimize parser, how dynamic CSS works
- Medium post *"Rockeyâ€Šâ€”â€Štips and tricks"*. There are too lot of tips and tricks that I want to share with you
- *"Components kit"*â€Šâ€”â€Šlibrary with easiest way to develop React components using rockey and [recompose](https://github.com/acdlite/recompose)
