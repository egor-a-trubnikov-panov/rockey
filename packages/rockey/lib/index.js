import condition from './condition';
import rule from './rule';
import props from './props';
import insert from './insert';
import when from './when';

const rockey = {
  condition,
  rule,
  insert,
  when,
};

export default rockey;
export { condition };
export { insert };
export { when };
export { rule };
export { props };
