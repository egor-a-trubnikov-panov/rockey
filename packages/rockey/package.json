{
  "name": "rockey",
  "version": "0.0.5",
  "description": "",
  "main": "index.js",
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "classnames": "^2.2.5",
    "lodash": "^4.17.4"
  },
  "files": [
    "index.js",
    "condition.js",
    "rule.js",
    "insert.js",
    "when.js",
    "rockey.min.js",
    "css",
    "mixins",
    "styleSheets",
    "utils"
  ],
  "scripts": {
    "prepublish": "node tasks/prepublish.js",
    "dev": "node tasks/prepublish.js --watch"
  }
}
