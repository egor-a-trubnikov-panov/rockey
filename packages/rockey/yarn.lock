# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


asap@^2.0.5:
  version "2.0.5"
  resolved "https://registry.yarnpkg.com/asap/-/asap-2.0.5.tgz#522765b50c3510490e52d7dcfe085ef9ba96958f"

classnames@^2.2.5:
  version "2.2.5"
  resolved "https://registry.yarnpkg.com/classnames/-/classnames-2.2.5.tgz#fb3801d453467649ef3603c7d61a02bd129bde6d"

css-vendor@^0.3.8:
  version "0.3.8"
  resolved "https://registry.yarnpkg.com/css-vendor/-/css-vendor-0.3.8.tgz#6421cfd3034ce664fe7673972fd0119fc28941fa"
  dependencies:
    is-in-browser "^1.0.2"

is-in-browser@^1.0.2:
  version "1.1.3"
  resolved "https://registry.yarnpkg.com/is-in-browser/-/is-in-browser-1.1.3.tgz#56ff4db683a078c6082eb95dad7dc62e1d04f835"

lodash@^4.17.4:
  version "4.17.4"
  resolved "https://registry.yarnpkg.com/lodash/-/lodash-4.17.4.tgz#78203a4d1c328ae1d86dca6460e369b57f4055ae"
