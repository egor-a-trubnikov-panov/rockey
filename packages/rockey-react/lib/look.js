import rule from 'rockey/rule';
import when from 'rockey/when';
import insert from 'rockey/insert';
import condition from 'rockey/condition';

import htmlTags from './htmlTags';

import rockey from './';

const createRockeyHoc = (BaseComponent, displayName, css) => {
  if (BaseComponent.extends) {
    return BaseComponent.extends(displayName, css);
  } else {
    return rockey(displayName, BaseComponent, css);
  }
};

const look = (
  BaseComponent,
  {
    extendBase = true,
  } = {}
) =>
  (...args) => {
    const css = rule(...args);

    return css.transform((tree, create) => {
      const components = Object.keys(tree.components);
      const size = components.length;

      if (!size) {
        throw new Error(
          'Rockey look should containt at least one root component'
        );
      }

      const parentDisplayName = components[0];
      const baseCss = create({
        components: {
          [parentDisplayName]: tree.components[parentDisplayName],
        },
      });

      const children = {};

      for (let i = 1; i < size; i++) {
        const displayName = components[i];

        const comopnentCss = create({
          components: {
            [displayName]: tree.components[displayName],
          },
        });

        if (extendBase) {
          comopnentCss.addParent(baseCss);
        }

        const ChildRockeyHoc = createRockeyHoc(
          BaseComponent,
          displayName,
          comopnentCss
        );

        children[displayName] = ChildRockeyHoc;
        BaseComponent[displayName] = ChildRockeyHoc;
      }

      const ParentRockeyHoc = createRockeyHoc(
        BaseComponent,
        parentDisplayName,
        baseCss
      );

      BaseComponent[parentDisplayName] = ParentRockeyHoc;

      return {
        ...children,
        [parentDisplayName]: ParentRockeyHoc,
      };
    });
  };

look.when = when;
look.condition = condition;
look.insert = insert;

for (const tag of htmlTags) {
  // ---- tag hoc lazy creation
  Object.defineProperty(look, tag, {
    get: () => (...args) => look(tag)(...args),
  });
}

export default look;
