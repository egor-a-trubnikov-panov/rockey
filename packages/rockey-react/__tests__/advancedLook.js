import React from 'react';
import 'react-dom';
import renderer from 'react-test-renderer';

import rockey from '../lib/';
import look from '../lib/look';

jest.mock('rockey/utils/hash', () => {
  return () => '{{ hash }}';
});

test('Advanced look #1', () => {
  const Button = rockey.button('MyButton');

  Button.look`
    Primary {
      color: blue;
    }

    Warning {
      color: red;
    }
  `;

  const { SuccessButton, AlertButton } = look(Button.Primary)`
    SuccessButton {
      border: 1px solid green;
    }

    AlertButton {
      border: 1px solid red;
    }
  `;

  const PrimaryTree = renderer
    .create(<Button.Primary>Button.Primary</Button.Primary>)
    .toJSON();
  expect(PrimaryTree).toMatchSnapshot();

  const WarningTree = renderer
    .create(<Button.Warning>Button.Warning</Button.Warning>)
    .toJSON();
  expect(WarningTree).toMatchSnapshot();

  const SuccessButtonTree = renderer
    .create(<SuccessButton>SuccessButton</SuccessButton>)
    .toJSON();
  expect(SuccessButtonTree).toMatchSnapshot();

  // expect(SuccessButtonTree).toEqual(
  //   renderer
  //     .create(
  //       <Button.Primary.SuccessButton>
  //         SuccessButton
  //       </Button.Primary.SuccessButton>
  //     )
  //     .toJSON()
  // );

  const AlertButtonTree = renderer
    .create(<AlertButton>AlertButton</AlertButton>)
    .toJSON();
  expect(AlertButtonTree).toMatchSnapshot();

  AlertButton.look`
    Large {
      font-size: 15px;
    }

    Small {
      font-size: 15px;
    }
  `;

  const AlertButtonLargeTree = renderer
    .create(<AlertButton.Large>AlertButton.Large</AlertButton.Large>)
    .toJSON();
  expect(AlertButtonLargeTree).toMatchSnapshot();

  const AlertButtonSmallTree = renderer
    .create(<AlertButton.Small>AlertButton.Small</AlertButton.Small>)
    .toJSON();
  expect(AlertButtonSmallTree).toMatchSnapshot();
});
