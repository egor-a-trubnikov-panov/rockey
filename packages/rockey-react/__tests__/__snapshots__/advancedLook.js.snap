// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Advanced look #1 1`] = `
<button
  className="Primary-{{ hash }} MyButton-{{ hash }}"
>
  Button.Primary
</button>
`;

exports[`Advanced look #1 2`] = `
<button
  className="Warning-{{ hash }} MyButton-{{ hash }}"
>
  Button.Warning
</button>
`;

exports[`Advanced look #1 3`] = `
<button
  className="SuccessButton-{{ hash }} Primary-{{ hash }} MyButton-{{ hash }}"
>
  SuccessButton
</button>
`;

exports[`Advanced look #1 4`] = `
<button
  className="AlertButton-{{ hash }} Primary-{{ hash }} MyButton-{{ hash }}"
>
  AlertButton
</button>
`;

exports[`Advanced look #1 5`] = `
<button
  className="Large-{{ hash }} AlertButton-{{ hash }} Primary-{{ hash }} MyButton-{{ hash }}"
>
  AlertButton.Large
</button>
`;

exports[`Advanced look #1 6`] = `
<button
  className="Small-{{ hash }} AlertButton-{{ hash }} Primary-{{ hash }} MyButton-{{ hash }}"
>
  AlertButton.Small
</button>
`;
